crucible-llvm/src/Lang/Crucible/LLVM/MemModel/MemLog.hs

```diff
@@ -75,23 +85,55 @@ import           Lang.Crucible.LLVM.MemModel.Pointer
 import           Lang.Crucible.LLVM.MemModel.Type
 import           Lang.Crucible.LLVM.MemModel.Value
 
+--------------------------------------------------------------------------------
+-- Allocations
+
 data AllocType = StackAlloc | HeapAlloc | GlobalAlloc
   deriving (Eq, Ord, Show)
 
 data Mutability = Mutable | Immutable
   deriving (Eq, Ord, Show)
 
+-- | Details of a single allocation. The @Maybe SymBV@ argument is either a
+-- size or @Nothing@ representing an unbounded allocation. The 'Mutability'
+-- indicates whether the region is read-only. The 'String' contains source
+-- location information for use in error messages.
+data AllocInfo sym =
+  forall w. (1 <= w) => AllocInfo AllocType (Maybe (SymBV sym w)) Mutability Alignment String
+
 -- | Stores writeable memory allocations.
 data MemAlloc sym
-     -- | Allocation with given block ID. The @Maybe SymBV@ argument is either a
-     -- size or @Nothing@ representing an unbounded allocation. The 'Mutability'
-     -- indicates whether the region is read-only. The 'String' contains source
-     -- location information for use in error messages.
-   = forall w. (1 <= w) => Alloc AllocType Natural (Maybe (SymBV sym w)) Mutability Alignment String
-     -- | Freeing of the given block ID.
-   | MemFree (SymNat sym) String
-     -- | The merger of two allocations.
-   | AllocMerge (Pred sym) [MemAlloc sym] [MemAlloc sym]
+    -- | A collection of consecutive basic allocations.
+  = Allocations (Map Natural (AllocInfo sym))
+    -- | Freeing of the given block ID.
+  | MemFree (SymNat sym) String
+    -- | The merger of two allocations.
+  | AllocMerge (Pred sym) (MemAllocs sym) (MemAllocs sym)
+
+-- | Memory allocations are represented as a list with the invariant
+-- that any two adjacent 'Allocations' constructors must be merged
+-- together.
+newtype MemAllocs sym = MemAllocs [MemAlloc sym]
+
+instance Semigroup (MemAllocs sym) where
+  (MemAllocs lhs_allocs) <> (MemAllocs rhs_allocs)
+    | Just (lhs_head_allocs, Allocations lhs_m) <- List.unsnoc lhs_allocs
```

Comment:
If this semigroup operation is called a lot, it may be worth optimizing access to the tail of the list by using some double-sided data structure, like `Data.Sequence`.

Reply:
In practice, I believe this semigroup operation is really only used to cons new operations onto the front, so a list probably isn't too bad. Changing the datatype would also require changing all the functions that traverse `MemAllocs`, which currently use pattern matching on the list constructors. On the other hand, making that traversal code more abstract wouldn't be a bad thing.

In either case, if we switch `MemAllocs` to use `Data.Sequence`, we should probably do `MemWrites` as well.

Reply:
I'd be in favor, generally, of making those traversals and such more abstract to provide us with more flexibility in the future with representation choices.  I don't know if it should be part of this PR or not.

Reply:
Revision 151e1f9b1cf2ea1642cc890c5be9880ebde4a1cb makes `MemAllocs` into an abstract datatype, so now all the traversals are done using the abstract interface provided by `Lang.Crucible.LLVM.MemModel.MemLog`.